### ISSUE TITLE
`ImplementationTypeSelector` no longer appears to support chaining calls to `AddClasses`

### ISSUE THREAD

**Issue by mpetito** [khellang/Scrutor#36]
In the docs, the following example shows how to scan for multiple types:

**Comment by khellang** [comment]
Ouch! Yeah, that looks like a regression. I've pushed a fixed version, v2.1.1, to NuGet and updated the tests to include some chained calls to `AddClasses`. Could you try it out and report back if it fixes the problem?

**Comment by mpetito** [comment]
It seems like v2.1.1 has fixed the chaining issue but may have introduced a new issue?

In my project there are two different entry points for scanning for services, and they now appear to behave differently:
 - application startup, which scans for types within one or more assembles
    `services.Scan(scan => ApplyTypeConventions(scan.FromAssembliesOf(...)));`
 - integration test startup, which scans for types within an integration test class
    `services.Scan(scan => ApplyTypeConventions(scan.AddTypes( [list of nested types from test class] )));`

Both call the same method to identify and register various types by convention:
```
void ApplyTypeConventions(IImplementationTypeSelector selector)
{
    selector.AddClasses(classes => classes.AssignableTo<TypeA>()).AsImplementedInterfaces().AsSelf().WithTransientLifetime();
    selector.AddClasses(classes => classes.AssignableTo<TypeB>()).AsImplementedInterfaces().WithTransientLifetime();
    ...
}
```

During application startup, where `ApplyTypeConventions` is passed `scan.FromAssembliesOf(...)` the services are identified and registered correctly.  However, during integration test startup, when it is passed a list of types via `scan.AddTypes(...)`, no services are registered.  In this case `classes` initially contains all of the passed types as expected (whether using chaining or not, so that seems fixed).

The `AddTypes` usage worked in v2.1.0, and both paths registered the correct types as long as chaining was not used in `ApplyTypeConventions`. Now regardless of chaining, the `AddTypes` usage fails.

**Comment by khellang** [comment]
Hmm. Yeah, that's an unfortunate scenario. The issue is that `FromAssembliesOf` and `AddTypes` are on two different "levels".

`AddTypes` implicitly also selects the given types as "implementation types", so there's really no need to do `AddClasses` as well.

This lets you do something like

```csharp
Collection.Scan(scan => scan
    .AddTypes<TransientService1, TransientService2>()
        .AsImplementedInterfaces()
        .WithSingletonLifetime());
```

Unfortunately, due to the chaining of the interfaces, this also allows you to do call `AddClasses`, which causes some confusion.

I can see that you'd expect this to work, so I've pushed a version, v2.1.2, that allows you to do this, but be aware; your code would first add the types passed to `AddTypes` (registered as self, transient), _then_ add the types selected in the `AddClasses` calls. This probably doesn't matter, but you should know that it happens.

**Comment by khellang** [comment]
Ideally, we'd have a `FromTypes` API that's on the same "level" as `FromAssembly` etc., so you could do

```csharp
Collection.Scan(scan => scan
    .FromTypes<TransientService1, TransientService2>()
        .AddClasses(classes => classes.AssignableTo<TransientService1>()).AsSelf().WithTransientLifetime()
        .AddClasses(classes => classes.AssignableTo<TransientService2>()).AsImplementedInterfaces().WithSingletonLifetime();
```

That would do exactly what you expect, but without implicitly adding `TransientService1` and `TransientService2` as self first.

**Comment by khellang** [comment]
> This makes sense, and part of my confusion may have been an expectation that `AddTypes` and `AddClasses` would behave differently.

Yeah, the naming is supposed to suggest that it does, in fact, add the types, without calling anything else. The problem is just that the returned interface includes the `AddClasses` methods, which suggests that you could chain the calls as you attempted.

> I'm not sure I understand the advantage of having `FromTypes` or `AddTypes` accept a couple of generic type arguments as opposed to just accepting a list of types.

There is no difference. The ones taking just a couple of generic arguments is just for convenience:

https://github.com/khellang/Scrutor/blob/7714611a40e2b0e14b862dffc7b9f95f6590cf61/src/Scrutor/TypeSelectorExtensions.cs#L5-L24

> If you did introduce a `FromTypes` method which accepted a list of types, then every other `From` method could be an extension method of that interface. This approach would also allow for project specific extension methods to be defined easily (e.g. `FromIntegrationTest` in my case, or `FromApplicationAssemblies` might be another common case).

Yes, that's the idea. I see value in both `AddTypes` and `FromTypes`, but having both might be even more confusing, even though they bring more consistency to the API. I might just obsolete the `AddTypes` methods and remove them in a future version.

**Comment by khellang** [comment]
@mpetito Did you get to try out v2.1.2? It's a bit of a hack (using both `FromAssembliesOf` and `AddTypes` at the same time), but I think it should fix your issue 😄

**Comment by mpetito** [comment]
I'm not sure I understand how I would use `FromAssembliesOf`.  I have a list of types that I want to scan; I don't want to scan all of the types from the assemblies corresponding to my list of types.

I agree that `AddTypes` is misleading and probably not necessary.  Even its description says "Will scan the types Type in types" which does not at all indicate that it will be *adding* the types implicitly as "implementation types".  Why would you even need Scrutor to do that? You could just call `services.AddTransient`.

**Comment by khellang** [comment]
> Why would you even need Scrutor to do that?

It was mainly added for unit testing, where you don't want to do actual scanning, but just provide a few types to register.

**Comment by khellang** [comment]
In my mind, I think the name `AddTypes` should make it clear that it actually adds the types. The only reason it is added as self is because you don't specify the type to register as, and self is the default.

**Comment by mpetito** [comment]
So how would I go about scanning for types from an `IEnumerable<Type>` in 2.1.2?

The following proposal as I understand it will not work as desired:

```
IEnumerable<Type> typesToConsider = ...
services.Scan(scan => scan.FromAssembliesOf(typesToConsider).AddClasses(...) ...)
```

Would this not scan all of the types contained in the assemblies of the types to consider?

**Comment by khellang** [comment]
> So how would I go about scanning for types from an `IEnumerable<Type>` in 2.1.2?

Exactly the same way you've always have:

https://github.com/khellang/Scrutor/blob/7714611a40e2b0e14b862dffc7b9f95f6590cf61/test/Scrutor.Tests/ScanningTests.cs#L16-L19

The issue is that you used both `AddTypes` and `FromAssembliesOf` the same way, which unfortunately didn't work very well. There's no need for `AddClasses` when you call `AddTypes`.

The code you pasted above (https://github.com/khellang/Scrutor/issues/36#issuecomment-325134790) should work in 2.1.2.

> Would this not scan all of the types contained in the assemblies of the types to consider?

Yes, it would 😄

**Comment by mpetito** [comment]
It seems like this is working. I'll do some further testing.
